### PR TITLE
[move-vm] Fixes to enum type implementation

### PR DESCRIPTION
**Pull request by wrwg**
## Description

Contains a number of fixes to the implementation of enum types in the VM

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

New test added, existing test caught new checks and need to be adapted

### CODE REVIEW

**Comment by wrwg** [comment]
* **#14657** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14657?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wrwg and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: third_party/move/move-bytecode-verifier/bytecode-verifier-tests/src/unit_tests/variant_name_test.rs
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/bytecode-verifier-tests/src/unit_tests/variant_name_test.rs:21]: Can we use a descriptive test name instead of comments like this? So that I don't need to decipher the compiler module. I have seen a few cases like this and in case we change/migrate tests it's extremely unhelpful. Just a few sentences would be nice as well.

File: third_party/move/move-bytecode-verifier/src/verifier.rs
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/verifier.rs:70]: Probably fine for this PR, but this is quite a bad pattern I think:
1) pub functions for tests can be just `cfg(any(test, features = "testing"))` to avoid misuse (though name is descriptive so it's okay).
2) there are 2 functions to verify the module. I would prefer to see 1 where the caller provides the config, otherwise these defaults are really error-prone.

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/check_duplication.rs:330]: Can we add a test for this?

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/check_duplication.rs:208]: Can we have tests for these?

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/framework/src/module_metadata.rs
Comment (ziaptos) [inline aptos-move/framework/src/module_metadata.rs:640]: Are these also double checked? Is that an issue?

**Review by vgao1996 — COMMENTED** [review]
Regarding the signature checker v2 changes:
1. It's all pesudo-metering and does not result in real gas charges
2. Is my understanding correct that enums are already featured gated elsewhere and have not been enabled on mainnet? If so, no need for extra feature-gating here, but if that's not the case, then feature gating is needed.

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (ziaptos) [inline third_party/move/move-bytecode-verifier/src/check_duplication.rs:242]: Nit: code duplicated 5 times. Maybe borderline small to stay like this. Your call.

File: third_party/move/move-binary-format/src/check_complexity.rs
Comment (ziaptos) [inline third_party/move/move-binary-format/src/check_complexity.rs:247]: Wouldn't this introduce replay issues?

**Review by ziaptos — COMMENTED** [review]
Just comments, will let Victor do the final stamp

File: third_party/move/move-bytecode-verifier/bytecode-verifier-tests/src/unit_tests/variant_name_test.rs
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/bytecode-verifier-tests/src/unit_tests/variant_name_test.rs:21]: Done

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/check_duplication.rs:208]: There is an open bug for creating a better test framework here. Its very difficult and not well maintainable to add tests for 'invalid' bytecode, because the compiler does only produce valid ones. A lot of features like this are untested today. However, this is just boilerplate of existing code, so I'm not too concerned in this case.

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/check_duplication.rs:242]: To do this consistently, I'd need to rewrite most code in this module, which is too much for this kind of quick fix.

File: third_party/move/move-bytecode-verifier/src/verifier.rs
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/verifier.rs:70]: Yes, not part of this PR, which should be focused and minimal.

File: third_party/move/move-bytecode-verifier/src/check_duplication.rs
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/check_duplication.rs:330]: See above. This PR will not boil the ocean to add tests where there weren't much before.

File: aptos-move/framework/src/module_metadata.rs
Comment (wrwg) [reply aptos-move/framework/src/module_metadata.rs:640]: This appears to be the intended design here. No double checking as those are different `.name` if this you refer to.

File: third_party/move/move-binary-format/src/check_complexity.rs
Comment (wrwg) [reply third_party/move/move-binary-format/src/check_complexity.rs:247]: As discussed on slack, no, because its not gas related.

**Comment by wrwg** [comment]
As discussed on slack, they aren't on mainnet, and moreover replay for tesnet passed.

**Review by vgao1996 — APPROVED** [review]

**Review by ziaptos — APPROVED** [review]
I'm OK stamping this given Slack conversation and other reviews

**Review by runtian-zhou — APPROVED** [review]
Stamp to unblock. In terms of testing I think we need three different tests:
1. Index out of bound for variants.
2. Complexity check for handles.
3. Generic check for variants.

Seems that we only have one tests? I think it would be great if we can add checks for all those three cases.

**Pull request merged**
